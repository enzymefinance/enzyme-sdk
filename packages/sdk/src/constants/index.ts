export * from "./misc.js";
export * from "./percentages.js";
export * from "./selectors.js";
export * from "./time.js";
