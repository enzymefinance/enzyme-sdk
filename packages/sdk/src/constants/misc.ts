export const SHARES_UNIT = 1_000_000_000_000_000_000n;
export const MAX_UINT_128 = 340_282_366_920_938_463_463_374_607_431_768_211_455n;
export const MAX_UINT_256 =
  115_792_089_237_316_195_423_570_985_008_687_907_853_269_984_665_640_564_039_457_584_007_913_129_639_935n;
export const ETH_ADDRESS = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
export const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
// Some tokens, e.g. vault shares, do not allow transfers to the zero address.
export const ZERO_ADDRESS_ALT = "0x0000000000000000000000000000000000000001";
export const SPECIFIC_ASSET_REDEMPTION_DUMMY_FORFEIT_ADDRESS = "0x000000000000000000000000000000000000aaaa";
export const LIB_INIT_GENERIC_DUMMY_ADDRESS = "0x0000000000000000000000000000000000009999";
