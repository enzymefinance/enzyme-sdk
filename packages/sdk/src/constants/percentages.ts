// BPS
export const ONE_PERCENT_IN_BPS = 100n;
export const FIVE_PERCENT_IN_BPS = 500n;
export const TEN_PERCENT_IN_BPS = 1_000n;
export const ONE_HUNDRED_PERCENT_IN_BPS = 10_000n;

// WEI
export const ONE_ONE_HUNDREDTH_PERCENT_IN_WEI = 100_000_000_000_000n;
export const ONE_PERCENT_IN_WEI = 10_000_000_000_000_000n;
export const FIVE_PERCENT_IN_WEI = 50_000_000_000_000_000n;
export const TEN_PERCENT_IN_WEI = 100_000_000_000_000_000n;
export const ONE_HUNDRED_PERCENT_IN_WEI = 1_000_000_000_000_000_000n;
