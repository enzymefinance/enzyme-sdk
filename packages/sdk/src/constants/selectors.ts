export const AAVE_V2_CLAIM_REWARDS_TO_SELF_SELECTOR = "0x41485304"; //"claimRewardsToSelf(address[], uint256)
export const REGISTRY_ADD_TO_LIST_SELECTOR = "0x8da3d736"; // addToList(uint256,address[])
export const REGISTRY_ATTEST_LISTS_SELECTOR = "0xb426f946"; // attestLists(uint256[],string[])
export const REGISTRY_CREATE_LIST_SELECTOR = "0xbe68406e"; // createList(address,uint8,address[])
export const REGISTRY_REMOVE_FROM_LIST_SELECTOR = "0x51d3a322"; // removeFromList(uint256,address[])
export const REGISTRY_SET_LIST_OWNER_SELECTOR = "0x956e8faa"; // setListOwner(uint256,address)
export const REGISTRY_SET_LIST_UPDATE_TYPE_SELECTOR = "0xee4483dc"; // setListUpdateType(uint256,uint8)
export const CURVE_MINTER_MINT_SELECTOR = "0x6a627842"; // mint(address)
export const CURVE_MINTER_MINT_MANY_SELECTOR = "0xa51e1904"; // mint_many(address[8])
export const CURVE_MINTER_TOGGLE_APPROVE_MINT_SELECTOR = "0xdd289d60"; // toggle_approve_mint(address)
export const PRICELESS_ASSET_BYPASS_START_ASSET_BYPASS_TIMELOCK_SELECTOR = "0x07d2890e"; // startAssetBypassTimelock(address)
export const SYNTHETIX_ASSIGN_EXCHANGE_DELEGATE_SELECTOR = "0x447fbc63"; // approveExchangeOnBehalf(address)
export const VAULT_CALL_ANY_DATA_HASH = "0x5bf1898dd28c4d29f33c4c1bb9b8a7e2f6322847d70be63e8f89de024d08a669"; // mln.vaultCall.any
