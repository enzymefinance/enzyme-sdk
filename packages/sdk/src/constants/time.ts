export const ONE_HOUR_IN_SECONDS = 3_600n;
export const ONE_DAY_IN_SECONDS = 86_400n;
export const ONE_WEEK_IN_SECONDS = 604_800n;
export const ONE_YEAR_IN_SECONDS = 31_557_600n;
