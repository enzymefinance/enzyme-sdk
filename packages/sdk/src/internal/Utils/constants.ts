export const SHARES_UNIT = 1_000_000_000_000_000_000n;
export const ETH_ADDRESS = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
export const ONE_HOUR_IN_SECONDS = 60n * 60n;
export const ONE_DAY_IN_SECONDS = ONE_HOUR_IN_SECONDS * 24n;
export const ONE_WEEK_IN_SECONDS = ONE_DAY_IN_SECONDS * 7n;
export const ONE_YEAR_IN_SECONDS = (ONE_DAY_IN_SECONDS * 36525n) / 100n;
export const ONE_PERCENT_IN_BPS = 100n;
export const FIVE_PERCENT_IN_BPS = ONE_PERCENT_IN_BPS * 5n;
export const TEN_PERCENT_IN_BPS = ONE_PERCENT_IN_BPS * 10n;
export const ONE_HUNDRED_PERCENT_IN_BPS = 10_000n;
export const ONE_HUNDRED_PERCENT_IN_WEI = 1_000_000_000_000_000_000n;
export const TEN_PERCENT_IN_WEI = ONE_HUNDRED_PERCENT_IN_WEI / 10n;
export const FIVE_PERCENT_IN_WEI = ONE_HUNDRED_PERCENT_IN_WEI / 20n;
export const ONE_PERCENT_IN_WEI = ONE_HUNDRED_PERCENT_IN_WEI / 100n;
export const ONE_ONE_HUNDREDTH_PERCENT_IN_WEI = ONE_HUNDRED_PERCENT_IN_WEI / 10_000n;
export const SPECIFIC_ASSET_REDEMPTION_DUMMY_FORFEIT_ADDRESS = "0x000000000000000000000000000000000000aaaa";
export const LIB_INIT_GENERIC_DUMMY_ADDRESS = "0x0000000000000000000000000000000000009999";
export const NULL_ADDRESS_ALT = "0x0000000000000000000000000000000000000001";
