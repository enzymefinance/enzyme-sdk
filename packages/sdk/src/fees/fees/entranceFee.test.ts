import { test, expect } from "vitest";
import {
  decodeEntranceRateDirectFeeSettings,
  decodeEntranceRateBurnFeeSettings,
  encodeEntranceRateBurnFeeSettings,
  encodeEntranceRateDirectFeeSettings,
  calculateEntranceRateFeeSharesDue,
  entraceRateBurnFeeSettingsEncoding,
  entraceRateDirectFeeSettingsEncoding,
} from "./entranceFee.js";

import { toBps } from "../../index.js";
import { vitalik } from "../../../tests/utils/constants.js";

test("should encode entrance rate burn fee settings correctly", () => {
  expect(encodeEntranceRateBurnFeeSettings({ feeRateInBps: toBps(0.123) })).toMatchInlineSnapshot(
    '"0x00000000000000000000000000000000000000000000000000000000000004ce"',
  );
});

test("should decode entrance rate burn fee settings correctly", () => {
  expect(
    decodeEntranceRateBurnFeeSettings("0x00000000000000000000000000000000000000000000000000000000000004ce"),
  ).toMatchInlineSnapshot(`
    {
      "feeRateInBps": 1230n,
    }
  `);
});

test("should encode entrance rate direct fee settings correctly", () => {
  expect(
    encodeEntranceRateDirectFeeSettings({
      feeRateInBps: toBps(0.567),
    }),
  ).toMatchInlineSnapshot(
    '"0x00000000000000000000000000000000000000000000000000000000000016260000000000000000000000000000000000000000000000000000000000000000"',
  );

  expect(
    encodeEntranceRateDirectFeeSettings({ feeRateInBps: toBps(0.567), feeRecipient: vitalik }),
  ).toMatchInlineSnapshot(
    '"0x0000000000000000000000000000000000000000000000000000000000001626000000000000000000000000d8da6bf26964af9d7eed9e03e53415d37aa96045"',
  );
});

test("should decode entrance rate direct fee settings correctly", () => {
  expect(
    decodeEntranceRateDirectFeeSettings(
      "0x0000000000000000000000000000000000000000000000000000000000001626000000000000000000000000d8da6bf26964af9d7eed9e03e53415d37aa96045",
    ),
  ).toMatchInlineSnapshot(`
    {
      "feeRateInBps": 5670n,
      "feeRecipient": "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
    }
  `);
});

test("should calculate entrance fee shares due correctly", () => {
  expect(
    calculateEntranceRateFeeSharesDue({ feeRateInBps: toBps(0.12345), sharesBought: 1000000000000000000n }),
  ).toEqual(123400000000000000n);
});

test("entraceRateBurnFeeSettingsEncoding should have correct properties", () => {
  expect(entraceRateBurnFeeSettingsEncoding).toMatchInlineSnapshot(`
    [
      {
        "name": "feeRate",
        "type": "uint256",
      },
    ]
  `);
});

test("entraceRateDirectFeeSettingsEncoding should have correct properties", () => {
  expect(entraceRateDirectFeeSettingsEncoding).toMatchInlineSnapshot(`
    [
      {
        "name": "feeRate",
        "type": "uint256",
      },
      {
        "name": "feeRecipient",
        "type": "address",
      },
    ]
  `);
});
