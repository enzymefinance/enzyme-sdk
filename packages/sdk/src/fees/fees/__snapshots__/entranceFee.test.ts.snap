// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should decode entrance rate direct fee settings correctly 1`] = `
{
  "feeRateInBps": 5670n,
  "feeRecipient": "0x0000000000000000000000000000000000000000",
}
`;
